### PR TITLE
Registry: Allow using class-strings instead of magic strings

### PR DESCRIPTION
**Pull request by Art4**
As discussed with @jtojnar I finalized #737 for using class-strings instead of strings in `Registry` class. #737 can be closed.

It also create legacy replacements to provide BC. In future it should also possible to use an interface as `$type` and provide an implementing class as `$class`.

### CODE REVIEW

File: src/Registry.php [hunk 1]
```diff
@@ -156,21 +178,13 @@ public function get_class($type)
...
-        }
-
-        // Try again with the legacy type resolved.
-        $type = $this->legacyClassNames[$type];
-        if (!empty($this->classes[$type])) {
-            return $this->classes[$type];
-        }
-        if (!empty($this->default[$type])) {
-            return $this->default[$type];
+        if (!empty($this->legacy[$type])) {
```
Comment (jtojnar) [inline src/Registry.php]: This will no longer return properly registered class:

```php
class CustomSniffer extends SimplePie\Content\Type\Sniffer {}

$registry->register(SimplePie\Content\Type\Sniffer::class, CustomSniffer::class);
var_dump($registry->get_class(`Content_Type_Sniffer`) instanceof CustomSniffer); // returns false
```

File: tests/Unit/SimplePieTest.php [hunk 2]
```diff
@@ -264,11 +269,6 @@ public function testLegacyCallOfSetCacheClass()
             $this->expectError();
         }
 
-        $feed = new SimplePie();
```
Comment (jtojnar) [inline tests/Unit/SimplePieTest.php:267]: Why this unrelated change?

File: src/Registry.php [hunk 3]
```diff
@@ -53,41 +57,33 @@
...
+     * Default class mapping
      *
+     * Overriding classes *must* subclass these.
      *
-     * @var array<string, class-string>
+     * @var array<class-string, class-string>
      */
-    protected $default = [];
+    protected $default = [
+        Cache::class => Cache::class,
```
[the remark this answers is not in the document]
Comment (Art4) [reply src/Registry.php:67]: This will allow us to have entries with `interface` => `classes` in the future. I have this use case while working on PSR-18 support.

File: src/Registry.php [hunk 4]
```diff
@@ -112,24 +128,30 @@ class Registry
...
+
         if (!class_exists($class)) {
             return false;
         }
 
-        if (!((isset($this->legacyClassNames[$type]) && is_subclass_of($class, $type)) || (isset($this->default[$type]) && is_subclass_of($class, $this->default[$type])))) {
+        /** @var string */
+        $base_class = $this->default[$type] ?? $this->legacy[$type];
+
+        if (!@is_subclass_of($class, $base_class)) {
```
[the remark this answers is not in the document]
Comment (Art4) [reply src/Registry.php:168]: not sure. It was muted like this earlier. I plan to replace this with a `instanceof` check

[the remark this answers is not in the document]
Comment (Art4) [reply src/Registry.php]: Fixed.

File: src/Registry.php [hunk 5]
```diff
@@ -136,27 +185,38 @@ public function register($type, $class, $legacy = false)
...
+     * @template T
+     * @param class-string<T> $type
+     * @return class-string<T>|null
      */
     public function get_class($type)
     {
-        if (!empty($this->classes[$type])) {
-            return $this->classes[$type];
+        if (array_key_exists($type, $this->legacyTypes)) {
+            // trigger_error(sprintf('"%s"(): Using argument #1 ($type) with value "%s" is deprecated since SimplePie 1.8, use class-string "%s" instead.', __METHOD__, $type, $this->legacyTypes[$type]), \E_USER_DEPRECATED);
```
Comment (mblaney) [inline src/Registry.php:195]: Do you want to remove or uncomment this error?

Comment (Art4) [reply src/Registry.php:195]: I would uncomment the trigger_error in the last minor release.

Comment (mblaney) [reply src/Registry.php:195]: If I merge this PR before the release it will be part of 1.8, so in that case the error seems applicable? Up to you though.

Comment (Art4) [reply src/Registry.php:195]: For consistency I would keep it as comment.

Comment (mblaney) [reply src/Registry.php]: ok I will merge this now then, I will wait for you to update the change log to include this PR before creating the release.
